File: rust-2020/src/day01.rs
use crate::runner;

fn parse_input() -> Vec<i32> {
    let mut input = Vec::new();
    if let Ok(lines) = runner::read_lines("input/01") {
        for line in lines {
            if let Ok(ip) = line {
                let parsed = ip.parse::<i32>().unwrap();
                input.push(parsed);
            }
        }
    }
    input.sort();
    input
}

fn part1() -> i32 {
    let input = parse_input();
    for x in &input {
        let s = 2020 - x;
        if let Ok(_) = &input.binary_search(&s) {
            return s * x;
        }
    }
    0
}

fn part2() -> i32 {
    let input = parse_input();
    for x in &input {
        for y in &input {
            if x + y > 2020 {
                break;
            }
            let s = 2020 - (x + y);
            if let Ok(_) = &input.binary_search(&s) {
                return s * x * y;
            }
        }
    }
    0
}

pub fn run() {
    let res1 = part1();
    println!("{}", res1);
    let res2 = part2();
    println!("{}", res2);
}

File: 2020/c-sharp/Day01.cs
using System;
using System.Collections.Generic;
using System.Linq;


namespace AOC2020
{
    public class Day01 {
        public static void Run(string input) {
            int[] lines = input.Split("\n").Select(x => int.Parse(x)).ToArray();
            Array.Sort(lines);

            void Part1() {
                foreach (var line in lines) {
                    var index = Array.BinarySearch(lines, 2020 - line);
                    if (index > 0) {
                        var part1 = lines[index] * line;
                        Console.WriteLine(part1);
                        return;
                    }
                }
            };
            void Part2() {
                for(int i = 0; i < lines.Length; i++) {
                    for(int j = i+1; j < lines.Length; j++) {
                        var index = Array.BinarySearch(lines, 2020 - lines[i] - lines[j]);
                        if (index > 0) {
                            var part2 = lines[index] * lines[i] * lines[j];
                            Console.WriteLine(part2);
                            return;
                        }
                    }
                }
            };
            Part1();
            Part2();
        }
    }
}

